src/clean-core/hash.hh

```diff
@@ -4,20 +4,64 @@
 #include <clean-core/fwd.hh>
 #include <clean-core/typedefs.hh>
 
+#include <cstring>
+#include <type_traits>
+
 namespace cc
 {
 template <class T, class>
 struct hash
 {
     static_assert(always_false<T>, "no hash<> specialization found");
 
-    // hash_t operator()(T const& value) const noexcept { ... }
+    // [[nodiscard]] hash_t operator()(T const& value) const noexcept { ... }
 };
 
+
+// ============== hash_combine ==============
+
+constexpr inline hash_t hash_combine() { return 0x2a5114b5c6133408uLL; }
+constexpr inline hash_t hash_combine(hash_t a) { return a; }
+constexpr inline hash_t hash_combine(hash_t a, hash_t b) { return a * 6364136223846793005ULL + b + 0xda3e39cb94b95bdbULL; }
+
+template <class... Args>
+constexpr hash_t hash_combine(hash_t a, hash_t b, hash_t c, Args... rest)
+{
+    static_assert((std::is_same_v<Args, hash_t> && ...), "extra arguments need to be hash_t as well");
+    auto h = hash_combine(a, b);
+    h = hash_combine(h, c);
+    ((h = hash_combine(h, rest)), ...);
+    return h;
+}
+
+
+// ============== make_hash ==============
+
+template <class... Args>
+constexpr hash_t make_hash(Args const&... values)
+{
+    return cc::hash_combine(hash<Args>{}(values)...);
+}
+
+
+// ============== default specializations ==============
+
 template <class T>
-struct hash<T*>
+struct hash<T, std::enable_if_t<std::is_trivially_copyable_v<T> && std::has_unique_object_representations_v<T>>>
 {
-    [[nodiscard]] hash_t operator()(T* const& value) const noexcept { return hash_t(value); }
-};
+    [[nodiscard]] hash_t operator()(T const& value) const noexcept
+    {
+        enum
+        {
+            wcnt = (sizeof(value) + sizeof(hash_t) - 1) / sizeof(hash_t)
```

Comment:
Why not a constexpr variable?

Reply:
because I don't want to generate symbols but guarantee constexpr eval (though it probably doesn't matter much here)